### ISSUE TITLE
Merge with MoBIE

### ISSUE THREAD

**Issue by tischi** [embl-cba/plateviewer#34]
@constantinpape

I have a feeling that the "hack" that we are doing with Yannick's tomograms could get us quite far in terms of using MoBIE as PlateViewer: one would "just" have to save LabelMask images for the wells and sites and the corresponding tables.

What do you think?

**Comment by constantinpape** [comment]
In general, I think merging this more with MoBIE is a good idea.
But I would rather have a bit cleaner solution than putting everything into one big array.
Instead we could think of a new type of dataset for MoBIE that corresponds to such "array" datasets via affines.

Something like this:
```
{"arrayDataset":
  "fileList": ["/some/file.n5", ...],
  "transforms": [[1, 1, 1..], ...]}
```

What do you think?

**Comment by tischi** [comment]
Not sure...

1. this reminds me of the discussion we had with Stephan (Saalfeld) about putting several datasets into one .n5 container. You think this is related?
1. Do we always want to adjust color and contrast for all images in an array simultaneously?
1. Do we sometimes want to rearrange them on the fly? For example, I think it could be cool to, e.g., sort all tomograms according to what they show rather than just showing them in a random order in a grid. But maybe this is overly fancy and it is enough to just be able to color them and then navigate to them via the table?!

**Comment by constantinpape** [comment]
>     1. this reminds me of the discussion we had with Stephan (Saalfeld) about putting several datasets into one .n5 container. You think this is related?

Yes, it's somewhat related. Anyway, why I don't like having everything in one big array:
1. it involves copying the data one more time and produces a very large array
2. it becomes harder to open the single images / tomogroms with an other software.

>     2\. Do we always want to adjust color and contrast for all images in an array simultaneously?

That's a good question, I think ideally we would want to allow both, but I am not sure if there is a good way to allow this in the UI.

>     3\. Do we sometimes want to rearrange them on the fly? For example, I think it could be cool to, e.g., sort all tomograms according to what they show rather than just showing them in a random order in a grid. But maybe this is overly fancy and it is enough to just be able to color them and then navigate to them via the table?!

Yes, re-arranging would be quite cool, but I think in the beginning navigation via the table and bookmarks should be enough.


I think we should discuss the storage question about point 1. in person at some point, but let's maybe wait till Yannick has some feedback about the current version of the tomogram dataset.

**Comment by tischi** [comment]
I am also for storing them the images in an array separate for the reason you mentioned (easier to understand and open with other software).

Another question: should we use the label mask logic to make things clickable and link to the table?

I wonder now whether I could, in java, generate label mask images on the fly for an array image and like this reuse the code in MoBIE....in fact, I looked at the code in the plateViewer and it seems I do something like this already anyway:

```
	private void createImage( )
	{
		BiConsumer< Localizable, IntType > biConsumer = ( l, t ) ->
		{
			t.setInteger( ListItemsARGBConverter.OUT_OF_BOUNDS_ROW_INDEX );

			for ( T annotatedIntervalTableRow : tableRows )
			{
				final Interval interval = annotatedIntervalTableRow.getInterval();
				if ( interval == null ) return;

				if ( Intervals.contains( interval, l ) )
				{
					t.setInteger( annotatedIntervalTableRow.rowIndex() );
				}
			}
		};

		final FunctionRandomAccessible< IntType > randomAccessible =
				new FunctionRandomAccessible( 2, biConsumer, IntType::new );

```

In this code the `interval` for each image could come from your proposed `arrayDataset` transforms.
We would need a convention/mechanism for how a row in the table maps to an image in the `arrayDataset`. Maybe a `label_id` column as we had it would work? The good thing would be that we could probably reuse some logic that exists already.

**Comment by constantinpape** [comment]
> Another question: should we use the label mask logic to make things clickable and link to the table?
> 
> I wonder now whether I could, in java, generate label mask images on the fly for an array image and like this reuse the code in MoBIE

Yes, I think this sounds like a good solution (if you can generate the label mask on the fly).

> We would need a convention/mechanism for how a row in the table maps to an image in the `arrayDataset`. Maybe a `label_id` column as we had it would work?

We can easily associate an id with each image in the arrayDataset, and then use this id to identify rows in the table.
I would rather call it something else than `label_id` (because this implies an actual label mask), maybe `array_id` or `grid_id`?
But if it makes a big difference in terms of code re-usability we can also stick with `label_id`.


There is one more question if we "fully" want to merge plateViewer functionality and MoBIE: what about grouping of images into wells? For this we would somehow need tables that are associated with a set of images.
But we can also think about this later, just wanted to bring it up now.

**Comment by tischi** [comment]
@martinschorb @constantinpape 

I think for MoBIE with the tomograms another thing to consider is that the global coordinate system in which we currently place them is completely meaningless. On an HTM plate there is some meaning, because it is relevant to know, e.g., whether a certain well was at the edge of the plate and the upper left corner of the plate is a natural global (0,0) coordinate.

However, for just a collection of tomograms we maybe want to do something else than putting them on a random grid.
Also for storing bookmarks for such tomograms I think it would be better if the coordinates were relative to the upper left corner of the respective tomogram and not relative to some random global origin. The other natural (and very nice thing) to do would be to have an overview image where the tomograms actually were acquired and place them into this overview image (I did that with Martin in another project).

1. What do you think?
2. Should we open a separate issue for this in MoBIE (rather than discussing here in the plateviewer)?

**Comment by constantinpape** [comment]
> I think for MoBIE with the tomograms another thing to consider is that the global coordinate system in which we currently place them is completely meaningless. On an HTM plate there is some meaning, because it is relevant to know, e.g., whether a certain well was at the edge of the plate and the upper left corner of the plate is a natural global (0,0) coordinate.

Yes, that makes sense, but we would need to coordinate systems (one for the larger grid and one for the individual grid images, e.g. tomograms). I am not sure how much overview this is and how much it would complicate the bookmarks etc.

> However, for just a collection of tomograms we maybe want to do something else than putting them on a random grid.

Yes, that makes sense.


>     1. What do you think?
> 
>     2. Should we open a separate issue for this in MoBIE (rather than discussing here in the plateviewer)?

Completely up to you, I think this depends on what you think makes more sense from your perspective w.r.t. sharing code / less implementation effort.

**Comment by martinschorb** [comment]
Hi,

I think there are two approaches to this:

1. If meaningful physical coordinates exist (plate scan, Overview + highmag details/CLEM), it makes total sense to use them (also for bookmarks and every annotation). This is already perfectly implemented for multiple sources, source types etc. in BDV-Playground (including group-wise or source-wise contrast adjustments, LUTs etc.)
I think it would make total sense to use this as underlying framework and then out a layer of segmentations/annotations etc. on top. The transformation from pixel data to physical space is a single transform that can just be appended to every derived data set.

2. The other idea of displaying multiple (spatially independent) sources dynamically in space would require some thinking. Especially what axis/plane do you use for 3D data with similar major axis lengths (cubes). I don't have a good idea here. A gallery/grid arrangements makes sense for 2D data or tomos where `z<<x=y` but not much more.

I tend to prefer **1** whenever possible and then work with a smart system of bookmarks. I thus would not necessarily put too much thought in **2** except for maybe dynamic grid arrangements. 
Maybe it could be coupled to the 'group' feature of BDV. One could group sources and then open a new BDV window where all these sources are displayed regardless of their physical coordinates in a more or less regular grid whose positions could be calculated by some smart combination taking (possibly different) voxel and bounding box sizes into account.

**Comment by tischi** [comment]
100% agreed with all of the things both of you said!

> One could group sources and then open a new BDV window where all these sources are displayed regardless of their physical coordinates in a more or less regular grid whose positions could be calculated by some smart combination taking (possibly different) voxel and bounding box sizes into account.

Exactly something like this would be good I thought and I also agree that

> ...The other idea of displaying multiple (spatially independent) sources dynamically in space would require some thinking

...but I think it could be worth it. 
@nicoKiaru already did some cool stuff to dynamically display brain slices in bdv, if I remember correctly. 

I also agree with 

>  we would need two coordinate systems  ... not sure how much overview this is and how much it would complicate the bookmarks etc.

Bookmarks would probably need some additional tag specifying whether they belong to the coordinate system of a specific Source (e.g., one of the tomograms) rather than the global coordinate system. Sounds like some work, indeed, and I would probably only do it if we find a very minimal and clever implementation.

We could have a zoom about this including @nicoKiaru and also @joshmoore, if he wants, maybe next week. 
@joshmoore may be interested in terms of a specification for ome-zarr for image set that could potentially be displayed in a grid-like fashion.

**Comment by joshmoore** [comment]
Sorry, this slipped past. Happy to chat.

**Comment by NicoKiaru** [comment]
Happy to chat as well. The idea behind the tricks with the brain slices is, at some point, to wrap your Sources as a `TransformedSource`. This applies a 3d affine transform to a source. This 3d affine transform is mutable, so you can change it transiently to switch between different visualization modes.

**Comment by tischi** [comment]
@NicoKiaru @joshmoore @constantinpape @martinschorb 

Cool, it would be amazing to have a zoom call about storage and visualisation of "grid-like datasets".
https://doodle.com/poll/hcryretpzwe32ex3

cc @StephanPreibisch: if you have time, it would be great if you could join as what you do in BigStitcher is very relevant for this.

**Comment by tischi** [comment]
I think there are actually (at least) two scenarios:

1. many images that belong to one "dataset", where each image naturally has a place in some global coordinate system, e.g.,
    - images from htm plates
2. many images that belong to one "dataset", where each image **does not** naturally has a place in some global coordinate system, e.g.,
    - many tomograms acquired, e.g., at the same day, maybe even from same specimen. Here it turns out that one nice visualisation mode is to display all of them next to each other in some random order in a grid.

**Comment by tischi** [comment]
Hi all,
Let's do the zoom on Friday 15:00 then.
@NicoKiaru @joshmoore @constantinpape 
I'll send a zoom link around just before.
